java-checks/src/main/java/org/sonar/java/checks/security/AuthorizationsStrongDecisionsCheck.java

```diff
@@ -0,0 +1,120 @@
+
+  private static final String AUTHENTICATION = "org.springframework.security.core.Authentication";
+  private static final String JAVA_OBJECT = "java.lang.Object";
+
+  private static final MethodMatchers ACCESS_DECISION_VOTER_VOTE = MethodMatchers.create()
+    .ofSubTypes("org.springframework.security.access.AccessDecisionVoter")
+    .names("vote")
+    .addParametersMatcher(AUTHENTICATION, JAVA_OBJECT, "java.util.Collection")
+    .build();
+
+  private static final MethodMatchers PERMISSION_EVALUATOR_HAS_PERMISSION = MethodMatchers.create()
+    .ofSubTypes("org.springframework.security.access.PermissionEvaluator")
+    .names("hasPermission")
+    .addParametersMatcher(AUTHENTICATION, JAVA_OBJECT, JAVA_OBJECT)
+    .addParametersMatcher(AUTHENTICATION, "java.io.Serializable", "java.lang.String", JAVA_OBJECT)
+    .build();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.METHOD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodTree methodTree = (MethodTree) tree;
+    if (ACCESS_DECISION_VOTER_VOTE.matches(methodTree)) {
+      reportNoStrongDecision(methodTree,
+        e -> e.is(Tree.Kind.IDENTIFIER) && "ACCESS_DENIED".equals(((IdentifierTree) e).name()), "vote", "ACCESS_DENIED");
+    } else if (PERMISSION_EVALUATOR_HAS_PERMISSION.matches(methodTree)) {
+      reportNoStrongDecision(methodTree, e -> e.asConstant(Boolean.class).filter(Boolean.FALSE::equals).isPresent(), "hasPermission", "false");
+    }
+  }
+
+  private void reportNoStrongDecision(MethodTree methodTree, Predicate<ExpressionTree> isStrongDecision, String methodName, String strongDecision) {
+    ReturnStatementVisitor returnStatementVisitor = new ReturnStatementVisitor(isStrongDecision);
+    methodTree.accept(returnStatementVisitor);
+    if (!returnStatementVisitor.takesStrongDecision()) {
+      reportIssue(methodTree.simpleName(), String.format("\"%s\" method should return at least one time %s.", methodName, strongDecision));
+    }
+  }
+
+  private static class ReturnStatementVisitor extends BaseTreeVisitor {
+
+    private final Predicate<ExpressionTree> isStrongDecision;
+    private boolean containsComplexReturn = false;
+    private boolean takesStrongDecision = false;
+
+    ReturnStatementVisitor(Predicate<ExpressionTree> isStrongDecision) {
+      this.isStrongDecision = isStrongDecision;
+    }
+
+    public boolean takesStrongDecision() {
+      // We don't compute the return value when the expression is complex (method call, expression),
+      // and consider it as strong decision to avoid FP
+      return takesStrongDecision || containsComplexReturn;
+    }
+
+    @Override
+    public void visitReturnStatement(ReturnStatementTree tree) {
+      ExpressionTree expression = tree.expression();
+      if (expression != null && (expression instanceof LiteralTree || expression.is(Tree.Kind.IDENTIFIER))) {
```

Comment:
An identifier could be different from `ACCESS_GRANTED`, `ACCESS_ABSTAIN`, `ACCESS_DENIED`, `true`, `false` like a variable, and in this case it could be complex. This is a FP:
```
    @Override
    public int vote(Authentication authentication, Object object, Collection collection) { // false-positive
      int access = ACCESS_DENIED;
      if(testAccess(authentication, object, collection)) {
        access = ACCESS_GRANTED;
      }
      return access;
    }
```